code/android-core-library/src/main/java/com/adobe/marketing/mobile/rulesengine/download/RulesDownloader.java

```diff
@@ -0,0 +1,271 @@
+     */
+    private final String cacheName;
+
+    private final Networking networkService;
+    private final CacheService cacheService;
+    private final DeviceInforming deviceInfoService;
+    private final RulesZipProcessingHelper rulesZipProcessingHelper;
+
+
+    public RulesDownloader(@NonNull final String cacheName,
+                           @NonNull final Networking networkService,
+                           @NonNull final CacheService cacheService,
+                           @NonNull final DeviceInforming deviceInfoService) {
+        this(cacheName, networkService, cacheService, deviceInfoService, new RulesZipProcessingHelper(deviceInfoService));
+
+    }
+
+    @VisibleForTesting
+    RulesDownloader(@NonNull final String cacheName,
+                    @NonNull final Networking networkService,
+                    @NonNull final CacheService cacheService,
+                    @NonNull final DeviceInforming deviceInfoService,
+                    @NonNull final RulesZipProcessingHelper rulesZipProcessingHelper) {
+
+        if (StringUtils.isNullOrEmpty(cacheName))
+            throw new IllegalArgumentException("Name cannot be null or empty");
+
+        this.cacheName = cacheName;
+        this.networkService = networkService;
+        this.cacheService = cacheService;
+        this.deviceInfoService = deviceInfoService;
+        this.rulesZipProcessingHelper = rulesZipProcessingHelper;
+    }
+
+    /**
+     * Loads rules from the {@code url} and invokes {@code callback} with the extracted rules.
+     * Additionally, the extracted content is cached in cache bucket with name {@code name} and {@code url} as the key
+     * in {@code CacheService}.
+     *
+     * @param url      the url from which the compressed rules are to be downloaded
+     * @param callback the callback that will be invoked with the result of the download
+     */
+    public void load(@NonNull final String url,
+                     @NonNull final RulesDownloadCallback callback) {
+        if (!UrlUtils.isValidUrl(url)) {
+            Log.trace(TAG, cacheName, "Provided download url: %s is null or empty. ", url);
+            callback.call(new RulesDownloadResult(null, RulesDownloadResult.Reason.INVALID_SOURCE));
+            return;
+        }
+
+        final CacheResult cacheResult = cacheService.get(cacheName, url);
+
+        final NetworkRequest networkRequest = new NetworkRequest(
+                url,
+                HttpMethod.GET,
+                null,
+                extractHeadersFromCache(cacheResult),
+                DEFAULT_CONNECTION_TIMEOUT_MS,
+                DEFAULT_READ_TIMEOUT_MS
+        );
+
+        final NetworkCallback networkCallback = response -> {
+            final RulesDownloadResult result = handleDownloadResponse(url, response);
+            callback.call(result);
+        };
+
+        networkService.connectAsync(networkRequest, networkCallback);
+    }
+
+    /**
```

Comment:
Can you also add `loadFromCache()` method to this class as it already knows about and handles caching?